substrate/frame/support/src/traits/try_runtime/mod.rs

```diff
@@ -145,29 +145,90 @@ impl core::str::FromStr for UpgradeCheckSelect {
 /// the storage items of your pallet.
 ///
 /// This hook should not alter any storage.
-pub trait TryState<BlockNumber> {
-	/// Execute the state checks.
-	fn try_state(_: BlockNumber, _: Select) -> Result<(), TryRuntimeError>;
+pub trait TryStateLogic<BlockNumber> {
+	fn try_state(_: BlockNumber) -> Result<(), TryRuntimeError>;
 }
 
 #[cfg_attr(all(not(feature = "tuples-96"), not(feature = "tuples-128")), impl_for_tuples(64))]
 #[cfg_attr(all(feature = "tuples-96", not(feature = "tuples-128")), impl_for_tuples(96))]
 #[cfg_attr(all(feature = "tuples-128"), impl_for_tuples(128))]
-impl<BlockNumber: Clone + sp_std::fmt::Debug + AtLeast32BitUnsigned> TryState<BlockNumber>
-	for Tuple
+impl<BlockNumber> TryStateLogic<BlockNumber> for Tuple
+where
+	BlockNumber: Clone + sp_std::fmt::Debug + AtLeast32BitUnsigned,
+{
+	fn try_state(n: BlockNumber) -> Result<(), TryRuntimeError> {
+		let mut errors = Vec::<TryRuntimeError>::new();
+
+		for_tuples!(#(
+			if let Err(err) = Tuple::try_state(n.clone()) {
+				errors.push(err);
+			}
+		)*);
+
+		Ok(())
+	}
+}
+
+/// Logic executed when `try-state` filters are provided in the `try-runtime` CLI.
+///
+/// Returning `true` for the provided ID will include this `try-state` logic in the overall tests
+/// performed.
+pub trait IdentifiableTryStateLogic<BlockNumber>: TryStateLogic<BlockNumber> {
+	fn matches_id(_id: &[u8]) -> bool;
+}
+
+#[cfg_attr(all(not(feature = "tuples-96"), not(feature = "tuples-128")), impl_for_tuples(64))]
+#[cfg_attr(all(feature = "tuples-96", not(feature = "tuples-128")), impl_for_tuples(96))]
+#[cfg_attr(all(feature = "tuples-128"), impl_for_tuples(128))]
+impl<BlockNumber> IdentifiableTryStateLogic<BlockNumber> for Tuple
+where
+	BlockNumber: Clone + sp_std::fmt::Debug + AtLeast32BitUnsigned,
+{
+	for_tuples!( where #( Tuple: TryStateLogic<BlockNumber> )* );
+	fn matches_id(id: &[u8]) -> bool {
+		for_tuples!( #( if Tuple::matches_id(id) { return true; } )* );
+		return false;
+	}
+}
+
+/// Execute some checks to ensure the internal state of a pallet is consistent.
+///
+/// Usually, these checks should check all of the invariants that are expected to be held on all of
+/// the storage items of your pallet.
+///
+/// This hook should not alter any storage.
+pub trait TryState<BlockNumber> {
+	/// Execute the state checks.
+	fn try_state(_: BlockNumber, _: Select) -> Result<(), TryRuntimeError>;
```

Comment:
Why is the block number passed in? It can just be read from the System pallet?

Reply:
I mean, this is just the `TryState` trait as it is today. I would not want to change this in this PR as well. Of course, if you think that we should change that here, I am open to do so. But not having the `BlockNumber` argument implicitly assumes all implementors always have a way to retrieve it. This is currently true for both pallet and runtime logic, but it's nevertheless one more assumption we would bake into this trait.